controllers/application_poller_controller.go

```diff
@@ -0,0 +1,278 @@
+			return
+		}
+
+		if err := p.apiClient.UpdateTemplate(ctx, templateURL, *template); err != nil {
+			p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to save experiment template in server", err), activity.URL)
+			return
+		}
+
+		p.log.Info("successfully completed resource scan", "activity", activity.URL)
+	case applications.TagRun:
+		// We wont compare existing scan with current scan
+		// so we can preserve changes via UI
+
+		// Get previous template
+		previousTemplate, err := p.apiClient.GetTemplate(ctx, templateURL)
+		if err != nil {
+			p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to get experiment template from server, a 'scan' task must be completed first", err), activity.URL)
+			return
+		}
+
+		// Overwrite current scan results with previous scan results
+		if err = server.APITemplateToClusterExperiment(exp, &previousTemplate); err != nil {
+			p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to convert experiment template", err), activity.URL)
+			return
+		}
+
+		// At this point the experiment should be good to create/deploy/run
+		// so let's create all the resources and #profit
+
+		// TODO
+		// try to clean up on failure ( might be a simple / blind p.client.Delete(ctx,generatedResources[i])
+		for i := range generatedResources {
+			// TODO generatedResource ( experiment ) does not contain the namespace
+			// not sure why yet
+			objKey, err := client.ObjectKeyFromObject(generatedResources[i])
+			if err != nil {
+				p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to get object key", err), activity.URL)
+				return
+			}
+
+			holder := &unstructured.Unstructured{}
+			holder.SetGroupVersionKind(generatedResources[i].GetObjectKind().GroupVersionKind())
+			err = p.client.Get(ctx, objKey, holder)
+			switch {
+			case apierrors.IsNotFound(err):
+				if err := p.client.Create(ctx, generatedResources[i]); err != nil {
+					p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to create object", err), activity.URL)
+					return
+				}
+			case err == nil:
+				// TODO This might need to be a patch instead of an update
+				// {"error":"failed to update object: experiments.optimize.stormforge.io \"01fcj078y60j74m11tnm7ga0yw-2a1d7d90\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update"}
+				if err := p.client.Update(ctx, generatedResources[i]); err != nil {
+					p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to update object", err), activity.URL)
+					return
+				}
+			default:
+				// Assume this should be a hard error
+				p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to get object", err), activity.URL)
+				return
+			}
+		}
+
+		p.log.Info("successfully created in cluster resources", "activity", activity.URL)
+	}
+}
+
+func (p *Poller) handleErrors(ctx context.Context, err error, activityURL string) {
+	failure := applications.ActivityFailure{FailureMessage: err.Error()}
```

Comment:
Still needs a reason or the failure never happened.

Reply:
Reasons ( scan failed / run failed )